### ISSUE TITLE
CI: Build Docker image in Travis CI and push it to update repository

### ISSUE THREAD

**Issue by mtreacy002** [anitab-org/mentorship-backend#523]
## Description
As a developer,
I need to speed up the development process by having ready to pull docker image from the AnitaB (upstream) repo. For this, I need to have Travis CI build a docker image when the commit being merged and push that new image to the upstream repository (docker hub and github),
so that I can run and test application simply by pulling that docker image from central repo and skip the process for setting up environment variables and installing dependencies. This also have future benefit in which my application is ready for scaling in case later I need to use/distribute the image horizontally across several network/servers.

## Mocks
N/A

References:

- [Why push your own image?](https://karlcode.owtelse.com/blog/2017/01/25/push-a-docker-image-to-personal-repository/)
- [dockerHub and Github integration](https://medium.com/better-programming/build-your-docker-images-automatically-when-you-push-on-github-18e80ece76af)
- [Patterns for Continuous Integration with docker on TravisCI](https://medium.com/mobileforgood/patterns-for-continuous-integration-with-docker-on-travis-ci-71857fff14c5)
- [When to avoid Docker](https://www.freecodecamp.org/news/7-cases-when-not-to-use-docker/)
- [Test github with docker](https://developer.okta.com/blog/2018/09/27/test-your-github-repositories-with-docker-in-five-minutes)
- [How to create, push and pull a docker image from DockerHub - youtube](https://www.youtube.com/watch?v=dU5112nqViY) 

## Acceptance Criteria

- [ ] The change does not break the system. This is tested by running the app  (outside docker) on both local virtual environment and remote servers.
- [ ] The change is tested by building and running docker locally to make sure the new built image can run the application with no issues.
- [ ] The change is tested by pushing the code change to own forked repository with own Travis CI and Codacy space and confirming the outcome of that PR passes Travis and Codacy tests. 
- [ ] The code change is then pushed to AnitaB upstream repository where the maintainers have made necessary adjustment to its own DockerHub and TravisCI (such as adding DockerHub credentials to TravisCi environment variables)
- [ ] The maintaners of the repo confirm that the image got pushed to the AnitaB DockerHub so it's ready to be pulled and used by members in the development process.
- [ ] A test is done by running the application on local machine using image that is pulled from the upstream repository.

## Definition of Done
- [ ] All of the required items are completed.
- [x] Approval by 1 mentor.

## Estimation
Few days (need collaboration with maintainers to test on upstream end)

**Comment by isabelcosta** [comment]
@mtreacy002 I don't quite understand the issue. Initially, I thought it was about running docker when deploying with TravisCI what do you mean by this?

**Comment by mtreacy002** [comment]
Sorry, @isabelcosta. You are right, it is about building docker with travisCI. The mentioning of scalability in the issue description is just one of the advantage of docker (which I wrote here to link with the conversation brought up on the [#make project production ready](https://anitab-org.zulipchat.com/#narrow/stream/222534-mentorship-system/topic/Make.20project.20production.20ready/near/192182554) stream on Zulip. I understand that we don't need to worry about this now, coz this is too far ahead (we need to get to production first 😀 ). But nothing wrong with laying out a strong foundation.

Regardless of its future use, to make the docker image as a source of version control after it got built in TravisCI, we need to store it somewhere so we can reuse it in the next build. If we only confirm that travis can build the image but not storing it centrally to be used in running our app, then this beats the purpose of the image being the sole source of version control (important for production). 

For development, it is useful for testing to have the option to build and run docker in our own local machine (as stated in the instruction on Mentorship-backend repository). The potential issue is there are many variables can cause the dependencies used in the image different from one machine to another. 
This issue is explain in [this article](https://karlcode.owtelse.com/blog/2017/01/25/push-a-docker-image-to-personal-repository/). If we have the centralised image that we can pull and run directly without having to build our own docker image first locally, that'll speed up the development process (skipping the process of setting up local environment, maybe?). 


To store and make use of the image build by travisCI so that we can have one centralised image from anitaB repo, we need to push to DockerHub after it got built. This way, the next TravisCI build will look for the latest image from this docker repository. We also need to link that DockerHub to our github repo as stated in the [article here](https://medium.com/better-programming/build-your-docker-images-automatically-when-you-push-on-github-18e80ece76af).

So, what I'm attempting to achieve here is a way for us to run our application simply by running the docker container pulled from anitaB repo instead of setting up our own local environment, as pointed out by [this article](https://developer.okta.com/blog/2018/09/27/test-your-github-repositories-with-docker-in-five-minutes) and in this [youtube video](https://www.youtube.com/watch?v=dU5112nqViY).

Along with the wonderful things we could achieve with docker, I feel the need to mention [an article](https://www.freecodecamp.org/news/7-cases-when-not-to-use-docker/) that mentioned the undesirable factors of using docker. Personally, considering our application is pretty complex, using docker as a version control is worth the trouble. 

I'm new in this, so my explanations here are only based on what I read from those articles (not on personal experience). To make this works, I need some support from everyone and mentors to give me some input whether or not I'm heading the right way. If not where to look to move forward  😉

**Comment by mtreacy002** [comment]
I've updated the issue description. Hopefully it's not as ambiguous on what it is aiming like it was before... ☺️

**Comment by mtreacy002** [comment]
Hi @isabelcosta. I have tested this using the image I built and pushed to my own docker hub repo last time (the one with automatic black runs from within docker) just for testing. I managed to pull the image and run the application on a totally new virtual environment that has no pip dependencies installed and inside a folder that has no mentorship-backend git been cloned 🎉 🎊 .

So, what I've done was as follows:
1. I moved outside of the project folder just to make sure I'm not running the application locally from the existing project root dir
(notice that I'm in the 'Devel' folder, the parent dir of my project folder.
<img width="638" alt="Screen Shot 2020-04-07 at 5 19 32 pm" src="https://user-images.githubusercontent.com/29667122/78641843-2584c400-78f5-11ea-9927-abb1561edcb3.png">


2. To make sure I do the test on a clean slate, I deleted the existing mentorship-backend docker container and docker image (notice the mentorship-backend was the first one on each list)

<img width="1371" alt="Screen Shot 2020-04-07 at 5 04 17 pm" src="https://user-images.githubusercontent.com/29667122/78640779-a6db5700-78f3-11ea-97dc-052a38f5e2e0.png">
<img width="1311" alt="Screen Shot 2020-04-07 at 5 04 47 pm" src="https://user-images.githubusercontent.com/29667122/78640870-c4a8bc00-78f3-11ea-8671-4b996f4d92b8.png">
<img width="1227" alt="Screen Shot 2020-04-07 at 5 05 14 pm" src="https://user-images.githubusercontent.com/29667122/78641337-73e59300-78f4-11ea-8938-7081307a6465.png">

2. Then I pulled the latest image I pushed to dockerhub last time I pushed to PR #511 (12 days ago)
<img width="768" alt="Screen Shot 2020-04-07 at 5 09 43 pm" src="https://user-images.githubusercontent.com/29667122/78642094-83191080-78f5-11ea-95e7-316bfa726379.png">
<img width="795" alt="Screen Shot 2020-04-07 at 5 06 16 pm" src="https://user-images.githubusercontent.com/29667122/78642194-ac39a100-78f5-11ea-832a-575adb3a2b52.png">

(this screenshot shows I have not installed pip dependencies in the new virtual environment and the latest image was pulled and now available to me locally)
<img width="1214" alt="Screen Shot 2020-04-07 at 5 06 52 pm" src="https://user-images.githubusercontent.com/29667122/78642367-f7ec4a80-78f5-11ea-8856-49e77be679c2.png">

3. then I just run the application on that unrelated folder and new virtual env by just borrowing .env file from the actual mentorship folder that I usually work on (I had to remove the word 'export' on each line of the .env to make this work).
<img width="1654" alt="Screen Shot 2020-04-07 at 5 09 13 pm" src="https://user-images.githubusercontent.com/29667122/78642825-ae502f80-78f6-11ea-8ca1-86dedd130354.png">

test register
<img width="1627" alt="Screen Shot 2020-04-07 at 5 54 05 pm" src="https://user-images.githubusercontent.com/29667122/78644381-23bcff80-78f9-11ea-95bd-f052bef542a7.png">

test login
<img width="1612" alt="Screen Shot 2020-04-07 at 5 55 33 pm" src="https://user-images.githubusercontent.com/29667122/78644434-36373900-78f9-11ea-8605-8d52acd453cb.png">


To summarise, a contributor can straight jump to working with the repository even without having to setup local environment. Of course, for us to modify the code and push to repo, we still need to actually work on the actual forked branch in our own local project root dir. This test was just to show I don't need to have anything else other than the modified .env and the docker image to run the application 😉.

Sorry again for the long details.... I got too excited that this actually works 😂 (I didn't expect it myself 😆 )

**Comment by mtreacy002** [comment]
Now I just need to learn how to debug with docker and trace where it had stored the data.... hehehe 😂

**Comment by mtreacy002** [comment]
Mmmm,....seems like the db needs to be dealt with differently. As soon as I stopped docker, the data also disappeared. I''ve used the postgreSql before with docker-compose but not sqlite-docker combo. Would this be an option we should explore (postgre-docker-compose)? Since I think we're also currently using postgre with the Heroku, aren't we? Pls let me know what you think, @isabelcosta 😉.

**Comment by vj-codes** [comment]
Heyy @mtreacy002 are you working on this as of now ?

**Comment by mtreacy002** [comment]
Hi @vj-codes. I was waiting for @isabelcosta to respond on the comment I posted above. But feel free to work on it coz I too want to know if there's a different way to approach this 😉. cc @isabelcosta

**Comment by vj-codes** [comment]
Ahhh no I was asking so as to label it as available or not🤦‍♀️ I'll remove the status available. Thanks! @mtreacy002

**Comment by SanketDG** [comment]
@mtreacy002 

If this issue is about setting up a docker development through Docker, I think we should not worry about database persistence. What we can do is to mention this in the docs that if you stop the container, you lose the data (even though this is implicit)

**Comment by epicadk** [comment]
> Now I just need to learn how to debug with docker and trace where it had stored the data.... hehehe 😂

Not sure about debugging but if you mount want the data that is stored in the docker container then you should read up on docker volumes.

**Comment by epicadk** [comment]
For debugging I found [this link](https://code.visualstudio.com/docs/containers/debug-common)

**Comment by mtreacy002** [comment]
closing this issue as it's no longer relevant to the project (TravisCI already moved to gh-action)
